src/libs/safe-path/src/safe_dir_builder.rs

```diff
@@ -0,0 +1,166 @@
+//
+
+use std::fs::DirBuilder;
+use std::io::{Error, ErrorKind, Result};
+use std::os::unix::fs::DirBuilderExt;
+use std::path::{Path, PathBuf};
+
+use crate::{safe_join, SafePathBuf};
+
+const DIRECTORY_MODE_DEFAULT: u32 = 0o700;
+const DIRECTORY_MODE_MASK: u32 = 0o777;
+
+/// Safe version of `DirBuilder` to protect from TOCTOU style of attacks.
+#[derive(Debug)]
+pub struct SafeDirBuilder {
+    root: PathBuf,
+    mode: u32,
+    recursive: bool,
+}
+
+impl SafeDirBuilder {
+    /// Creates a new set of options with default mode/security settings for all platforms and
+    /// also non-recursive.
+    pub fn new<P: AsRef<Path>>(root: P) -> Result<Self> {
+        let root = root.as_ref().canonicalize()?;
+        if !root.is_dir() {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", root.display()),
+            ));
+        }
+
+        Ok(SafeDirBuilder {
+            root,
+            mode: DIRECTORY_MODE_DEFAULT,
+            recursive: false,
+        })
+    }
+
+    /// Indicates that directories should be created recursively, creating all parent directories.
+    ///
+    /// Parents that do not exist are created with the same security and permissions settings.
+    pub fn recursive(&mut self) -> &mut Self {
+        self.recursive = true;
+        self
+    }
+
+    /// Sets the mode to create new directories with. This option defaults to 0o755.
+    pub fn mode(&mut self, mode: u32) -> &mut Self {
+        self.mode = mode & DIRECTORY_MODE_MASK;
+        self
+    }
+
+    /// Creates the specified directory with the options configured in this builder.
+    ///
+    /// The `path` must be a subdirectory of `SafePathBuf::root()`, otherwise error will be returned.
+    /// It is considered an error if the directory already exists unless recursive mode is enabled.
+    pub fn create<P: AsRef<Path>>(&self, path: P) -> Result<SafePathBuf> {
+        let mut root = self.root.clone();
+        let path = safe_join("/", path)?;
+        let mut suffix = path.strip_prefix(&root).map_err(|_| {
+            Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", path.display()),
+            )
+        })?;
+        if suffix.file_name().is_none() {
+            return SafePathBuf::from_path(root);
+        }
+        if !self.recursive {
+            if let Some(parent) = path.parent() {
+                root = root.join(parent);
+            }
+            // Safe to unwrap() because we have verified `suffix` is not empty.
+            suffix = Path::new(suffix.file_name().unwrap());
+        }
+
+        for comp in suffix {
+            let file = SafePathBuf::from_path(&root)?;
+            if !file.target().is_dir() {
+                return Err(Error::new(
+                    ErrorKind::Other,
+                    format!("Invalid path: {}", root.display()),
+                ));
+            }
+            root = root.join(comp);
```

Comment:
I believe this line throws away the safety you were trying to create.  The original `safe_join` made sure there were no symlinks in the suffix *at that time*.  But it obviously couldn't check components that didn't exist at the time.  If a symlink with the name of one of the suffix components was created between the `safe_join` and here, this could generate a path linking outside `root`.

I think to do this safely you're going to want to use `mkdirat()`.

Reply:
Yes, found the same issue when addressing earlier comments and have changed to use mkdirat().